limit the number of artifacts directories retained on disk

d-honeybadger opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #311 

This prevents too much data from accumulating on disk, especially now that retries are happening since each retry creates an artifacts directory.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->
1) Create an ansiblerun - a failing one is best cause it'll be retried without one having to change the spec
2) exec into the provider-ansible pod, navigate to `ansibleDir/$UID-of-the-run/artifacts`
3) watch that the number of directories never exceeds the limit (you might want to set a limit lower than default for speedy testing)

[contribution process]: https://git.io/fj2m9

d-honeybadger commented on internal/controller/ansibleRun/ansibleRun.go:
```diff
@@ -91,7 +91,7 @@ type ansibleRunner interface {
 
 // Setup adds a controller that reconciles AnsibleRun managed resources.
-func Setup(mgr ctrl.Manager, o controller.Options, ansibleCollectionsPath, ansibleRolesPath string, timeout time.Duration) error {
+func Setup(mgr ctrl.Manager, o controller.Options, ansibleCollectionsPath, ansibleRolesPath string, timeout time.Duration, artifactsHistoryLimit int) error {
```
Lmk if you'd like me to refactor so that `Setup`'s list of arguments doesn't keep growing forever

fahedouch replied:
yes please ^^

d-honeybadger replied:
Done! Since the config options for the `ansiblerun` controller setup and `config` controller setup keep growing apart, I decided to have a config/settings struct just for `ansiblerunner` (therefore the two `Setup` functions no longer have the same signature. I think it makes sense but lmk if you'd like to do something different

fahedouch replied:
thanks

fahedouch commented:
integration test is very welcome :)

fahedouch commented on cmd/provider/main.go:
```diff
@@ -42,6 +42,7 @@ func main() {
 		leaderElection         = app.Flag("leader-election", "Use leader election for the controller manager.").Short('l').Default("false").OverrideDefaultFromEnvar("LEADER_ELECTION").Bool()
 		maxReconcileRate       = app.Flag("max-reconcile-rate", "The maximum number of concurrent reconciliation operations.").Default("1").Int()
+		artifactsHistoryLimit  = app.Flag("artifacts-history-limit", "Each attempt to run the playbook/role generates a set of artifacts on disk. This settings limits how many of these to keep.").Default("50").Int()
 	)
```
IMO 50 is still huge

d-honeybadger replied:
That's fair - dropped to 10

d-honeybadger commented:
> integration test is very welcome :)

Since there's no prior art, are you thinking something like this: https://github.com/crossplane/crossplane/blob/master/test/e2e/README.md? e.g. run the provider against a `kind` k8s cluster and have ansible-runner actually run playbooks
Or I could add a test without `kind` cluster but with ansible-runner installed and running real playbooks - this would be more like a unit test but with real ansible, so I guess it'd just test the ansiblerun controller and not the entire flow of setting a flag

fahedouch commented:
> > integration test is very welcome :)
> 
> Since there's no prior art, are you thinking something like this: https://github.com/crossplane/crossplane/blob/master/test/e2e/README.md? e.g. run the provider against a `kind` k8s cluster and have ansible-runner actually run playbooks Or I could add a test without `kind` cluster but with ansible-runner installed and running real playbooks - this would be more like a unit test but with real ansible, so I guess it'd just test the ansiblerun controller and not the entire flow of setting a flag

I think it is fine to go with unit test for the moment.

let's introduce integration test in a follow-up PR! we can be inspired from other projects such as [aws provider](https://github.com/crossplane-contrib/provider-aws/blob/master/cluster/local/integration_tests.sh)

d-honeybadger commented:
> I think it is fine to go with unit test for the moment.

Added a couple unit tests. I know I mentioned in the beginning using the actual ansible-runner, but when I looked closer, there's no good way to do that without running the test in a container with a custom dockerfile: ansible-runner install requires python, and that's not something we can't just `curl` the binary for. Adding makefile targets and/or scripts that mess with python install on the system seemed like a bad idea, and doing a custom dockerfile and make targets to build that dockerfile seemed like an overkill too, especially since there's the plan to do e2e tests with `kind`.

So the tests I added are your typical unit tests using dummy commands, just checking that the runner settings and command args are set correctly.

fahedouch reviewed (approved):
Thanks
